### PR TITLE
Fix GPU memory transfer + EwaldInit race condition

### PR DESCRIPTION
**Pull request by spencerw**
This fixes a race condition where serializeLocal can sometimes be invoked before EwaldInit finishes. An EwaldMsg object is now passed to EwaldGPU, which tells whether EwaldInit, or the callback from the GPU memory transfer invoked it.

Note that the EwaldMsg only contains a single field here. EwaldMsg is used more extensively after the WorkRequest code has been removed from ChaNGa. These changes are contained in a separate pull request.

### CODE REVIEW

File: DataManager.cpp [hunk 1]
```diff
@@ -506,14 +505,14 @@ void DataManager::serializeLocalTree(){
...
 void DataManager::startLocalWalk() {
-    gputransfer = true;
     delete localTransferCallback;
     for(int i = 0; i < registeredTreePieces.length(); i++){
       if(verbosity > 1) CkPrintf("[%d] GravityLocal %d\n", CkMyPe(), i);
       int in = registeredTreePieces[i].treePiece->getIndex();
       treePieces[in].commenceCalculateGravityLocal();
       if(registeredTreePieces[0].treePiece->bEwald) {
-          dummyMsg *msg = new (8*sizeof(int)) dummyMsg;
+          EwaldMsg *msg = new EwaldMsg;
```
Comment (trquinn) [inline DataManager.cpp:514]: new needs an argument to specify the priority field (in bits).  E.g. "msg = new (8*sizeof(int)) EwaldMsg;"

File: Ewald.cpp [hunk 2]
```diff
@@ -373,9 +373,9 @@ void TreePiece::EwaldInit()
 			}
 		}
 	nEwhLoop = i;
-        bEwaldInited = true;
 
-	dummyMsg *msg = new (8*sizeof(int)) dummyMsg;
+	EwaldMsg *msg = new EwaldMsg;
```
Comment (trquinn) [inline Ewald.cpp:377]: See comment on similar line in DataManager.cpp

File: ParallelGravity.h [hunk 3]
```diff
@@ -325,6 +325,10 @@ struct BucketMsg : public CkMcastBaseMsg, public CMessage_BucketMsg {
   int whichTreePiece;
 };
 #endif
+
+struct EwaldMsg: public CkMcastBaseMsg, public CMessage_EwaldMsg {
```
Comment (trquinn) [inline ParallelGravity.h:329]: EwaldMsg does not need to be derived from CkMcastBaseMsg.

**Review by trquinn — CHANGES_REQUESTED** [review]
See code comments.  I'm still testing this out.

**Comment by spencerw** [comment]
I can confirm that the 'testcosmo' case passes after these changes have been made to the code.

**Review by trquinn — APPROVED** [review]

**Pull request merged**